File: vendas-ui/src/app/models/venda.ts
import { Item, ItemDTO } from "./item";
import { Cliente } from "./cliente";

export class Venda {
  id: string = '';
  cliente: Cliente = new Cliente();
  cadastro: string = '';
  frete: number = 0;
  itens: Item[] = [];
  total: number = 0;

  public toInsertDTO(): VendaInsertDTO {
    let itensDTO:ItemDTO[]  = [];
     this.itens.forEach(i => {
       itensDTO.push(new ItemDTO(i));
     });
    return {
      clienteId:this.cliente.id,
      frete:this.frete,
      itens:itensDTO
    };
  }
}

export class VendaInsertDTO {
  clienteId: string;
  frete: number;
  itens: ItemDTO[];

  // constructor(venda: Venda) {
  //   this.cliente = venda.cliente.id;
  //   this.frete = venda.frete;
  //   venda.itens.forEach(i => 
  //     this.itens.push(new ItemDTO(i))
  //   );
  // }
}

File: api-vendas/src/main/kotlin/br/com/juniorstreichan/apivendas/domain/Venda.kt
package br.com.juniorstreichan.apivendas.domain

import java.time.LocalDateTime
import javax.persistence.*
import javax.validation.Valid
import javax.validation.constraints.NotEmpty
import javax.validation.constraints.NotNull
import javax.validation.constraints.PositiveOrZero


@Entity
data class Venda(

        val cadastro: LocalDateTime = LocalDateTime.now(),

        @Id
        @GeneratedValue(strategy = GenerationType.IDENTITY)
        private var id: Int = 0,

        @ManyToOne
        private var cliente: Cliente? = null,

        @Valid
        @NotEmpty
        @OneToMany(mappedBy = "venda")
        val itens: MutableList<Item> = mutableListOf<Item>(),

        @PositiveOrZero
        private var frete: Double = 0.0,

        @PositiveOrZero
        private var total: Double = 0.0

) {
    fun setId(id: Int) = { this.id = id }

    fun getId(): Int = id

    fun getCliente(): Cliente? = cliente

    fun setCliente(cliente: Cliente) = { this.cliente = cliente }

    fun getFrete(): Double = frete

    fun setFrete(valor: Double) { frete = valor }

    fun getTotal():Double = total

    fun setTotal(valor:Double) {total = valor}


}